src/musig.rs

```diff
@@ -0,0 +1,1158 @@
+//! This module implements high-level Rust bindings for a Schnorr-based
+//! multi-signature scheme called MuSig2 [paper](https://eprint.iacr.org/2020/1261).
+//! It is compatible with bip-schnorr.
+//!
+//! The documentation in this module is for reference and may not be sufficient
+//! for advanced use-cases. A full description of the C API usage along with security considerations
+//! can be found in [C-musig.md](secp256k1-sys/depend/secp256k1/src/modules/musig/musig.md).
+use core;
+use core::fmt;
+use core::mem::MaybeUninit;
+#[cfg(feature = "std")]
+use std;
+
+use crate::ffi::{self, CPtr};
+use crate::{
+    schnorr, Error, Keypair, Message, PublicKey, Scalar, Secp256k1, SecretKey, Signing, Verification, XOnlyPublicKey
+};
+
+/// Musig partial signature parsing errors
+#[derive(Debug, Clone, Copy, Eq, PartialEq, PartialOrd, Ord, Hash)]
+pub enum ParseError {
+    /// Parse Argument is malformed. This might occur if the point is on the secp order,
+    /// or if the secp scalar is outside of group order
+    MalformedArg,
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for ParseError {}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        match *self {
+            ParseError::MalformedArg => write!(f, "Malformed parse argument"),
+        }
+    }
+}
+
+/// Session Id for a MuSig session.
+#[allow(missing_copy_implementations)]
+#[derive(Debug, Eq, PartialEq)]
+pub struct SessionSecretRand([u8; 32]);
+
+impl SessionSecretRand {
+    /// Generates a new session ID using thread RNG.
+    #[cfg(all(feature = "rand", feature = "std"))]
+    pub fn new() -> Self {
+        Self::from_rng(&mut rand::thread_rng())
```

Comment:
In dd26c3067044a89c503a32079c24227d6e1afca5:

Clippy is complaining about there being `new` but not `Default`. I think it's got a point, and that we should rename this function `from_thread_rng` to make it more obvious that it's not a pure function. Will do in a followup.